File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/authentication/AuthServiceImpl.java
package com.marknkamau.justjavastaff.authentication;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.marknkamau.justjavastaff.models.Employee;

/**
 * Created by Mark Njung'e.
 * mark.kamau@outlook.com
 * https://github.com/MarkNjunge
 */

public class AuthServiceImpl implements AuthenticationService {
    private FirebaseAuth auth;

    public AuthServiceImpl() {
        auth = FirebaseAuth.getInstance();
    }

    @Override
    public void signIn(String email, String password, final AuthListener listener) {
        auth.signInWithEmailAndPassword(email, password)
                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                    @Override
                    public void onSuccess(AuthResult authResult) {
                        listener.onSuccess(getEmployeeFromUser(authResult.getUser()));
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        listener.onError(e.getMessage());
                    }
                });
    }

    @Override
    @Nullable
    public Employee currentEmployee() {
        FirebaseUser currentUser = auth.getCurrentUser();
        if (currentUser == null ){
            return null;
        }else {
            return getEmployeeFromUser(currentUser);
        }
    }

    private Employee getEmployeeFromUser(FirebaseUser user){
        //TODO Admin interface should create users with a display name.
        String name = user.getEmail().split("@")[0];

        return new Employee(user.getUid(), name, user.getEmail());
    }
}

File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/authentication/AuthServiceImpl.kt
package com.marknkamau.justjavastaff.authentication

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.marknkamau.justjavastaff.models.Employee

/**
 * Created by Mark Njung'e.
 * mark.kamau@outlook.com
 * https://github.com/MarkNjunge
 */

class AuthServiceImpl : AuthenticationService {
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()

    override fun signIn(email: String, password: String, listener: AuthenticationService.AuthListener) {
        auth.signInWithEmailAndPassword(email, password)
                .addOnSuccessListener { authResult -> listener.onSuccess(getEmployeeFromUser(authResult.user)) }
                .addOnFailureListener { e -> listener.onError(e.message ?: "Error signing in") }
    }

    override fun signOut() {
        auth.signOut()
    }

    override fun currentEmployee(): Employee? {
        val currentUser = auth.currentUser
        return if (currentUser == null) {
            null
        } else {
            getEmployeeFromUser(currentUser)
        }
    }

    private fun getEmployeeFromUser(user: FirebaseUser): Employee {
        //TODO Admin interface should create users with a display name.
        val name = user.email!!.split("@".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]

        return Employee(user.uid, name, user.email!!)
    }
}
